CASSGO-13 Add support for cassandra 4.0 table options

OleksiienkoMykyta opened the pull request:
Cassandra 4.0 has some minor table metadata changes that need to be accounted for:
read_repair_chance is replaced with read_repair
dclocal_read_repair_chance has been removed
additional_write_policy has been added

In the PR implemented backward compatibility with previous versions, and added new types support. To make metadata table support easier for future Cassandra versions, hardcode scan from Cassandra were replaced with new "parseSystemSchemaViews" method which is much easier to expand, even if some fields were added in the middle of the table it wouldn`t be an issue anymore.

testisnullus commented on cassandra_test.go:
```diff
@@ -2278,12 +2283,17 @@ func TestMaterializedViewMetadata(t *testing.T) {
+		MemtableFlushPeriodInMs: 0,
+		MinIndexInterval:        128,
+		ReadRepair:              expectedReadRepair,
+		ReadRepairChance:        0,
```
As far as I know, read_repair_chance was removed from C* 4.0. 

https://issues.apache.org/jira/browse/CASSANDRA-15604

OleksiienkoMykyta replied:
Yes, but this test also runs for C* 3.0.0, so it covers data for all versions. We can make separate tests for each version, but as mentioned in the description for the current PR, the main aim is to implement a backward-compatible metadata getter, and this test case helps to test one.

worryg0d commented on metadata.go:
```diff
@@ -975,69 +977,207 @@ func getViewsMetadata(session *Session, keyspaceName string) ([]ViewMetadata, er
 	return views, nil
 }
 
+func extensionsBlobToStrings(value interface{}) (map[string]string, error) {
+	const errorMessage = "gocql.extensionsBlobToStrings failed to read column %s"
+	byteData, ok := value.(map[string][]byte)
+	if !ok {
+		return nil, fmt.Errorf(errorMessage, "extentions")
+	}
+
+	extensions := make(map[string]string, len(byteData))
+	for key, rowByte := range byteData {
+		extensions[key] = string(rowByte)
+	}
+
+	return extensions, nil
+}
```
```suggestion
func bytesMapToStringMap(bytesMap map[string][]byte) map[string]string {
	stringMap := make(map[string]string, len(bytesMap))
	for key, bytes := range bytesMap {
		stringMap[key] = string(bytes)
	}

	return stringMap
}
```

Let's make this method take `map[string][]byte` as a param instead of an empty interface. This approach will allow you to return described in this func error more properly from `materializedViewMetadataFromMap`.
Also, what do you think about renaming this func to something like `bytesMapToStringMap`?

OleksiienkoMykyta replied:
I agree, it will make bytesMapToStringMap more reusable and error handling much more readable.
Thank You very much!

joao-r-reis commented on metadata.go:
```diff
@@ -139,6 +140,7 @@ type MaterializedViewMetadata struct {
 	MaxIndexInterval        int
 	MemtableFlushPeriodInMs int
 	MinIndexInterval        int
+	ReadRepair              string
```
Maybe a small comment here and `ReadRepairChance` to indicate that `ReadRepairChance` was removed in C* 4.0

OleksiienkoMykyta replied:
Something like `ReadRepairChance        float64 // Deprecated since Cassandra v4, use ReadRepair instead.`?

joao-r-reis replied:
Maybe something like:

```
ReadRepair              string    // Only present in Cassandra 4.0+
ReadRepairChance        float64   // Note: Cassandra 4.0 removed ReadRepairChance and added ReadRepair instead
```

OleksiienkoMykyta committed:
Add support for cassandra 4.0 table options

In the PR implemented backward compatibility with previous versions,
and added new types support. To make metadata table support easier for
future Cassandra versions, hardcode scan from Cassandra were replaced
with new "parseSystemSchemaViews" method which is much easier to expand,
even if some fields were added in the middle of the table it wouldn`t be an issue anymore.

patch by Mykyta Oleksiienko; reviewed by Joao Reis and James Harting CASSGO-13

joao-r-reis commented:
Can you resolve the conflict and add James to the commit message?
